incentivization.md

```diff
@@ -0,0 +1,240 @@
+In what follows, we abbreviate incentivization as i13n.
+
+We aim to answer the following questions:
+
+1. what is the structure of the protocols in question?
+2. what is the desired behavior of protocol participants?
+3. what deviations from the desired behavior occur without incentivization?
+4. what incentivization tools do we have?
+5. what tools are appropriate for our purpose?
+6. what parameters can we chose? what are our restrictions?
+7. suggest a concrete i13n architecture.
+8. how do we check if we've solved the problem?
+
+# Overview
+
+Waku implements a modular decentralized censorship resistant P2P communications protocol.
+Waku consists of multiple protocols (see [architecture](https://waku.org/about/architect)).
+We focus on the main four are Relay (a P2P protocol), and three light protocols: Filter, Store, and Lightpush, which have a client-server architecture (aka request-response).
+
+A Waku node is a node that runs at least one of the Waku protocols.
+A full Waku node is a node that runs Relay.
+A light Waku node as a node that only runs client-side of one of the light protocols.
+See also: https://github.com/waku-org/research/issues/28
+
+In light protocols, a client sends a request to a server.
+A server (a Relay node) performs some actions and returns a response, in particular:
+- [[Filter]]: the server will relay (only) messages that pass a filter to the client;
+- [[Store]]: the server responds with messages broadcast within the specified time frame;
+- [[Lightpush]]: the server publishes the client's message to the Relay network.
+
+Waku aims to function on widely available hardware.
+Hardware requirements for light nodes are lower than for full nodes.
+In particular, bandwidth consumption should be limited (estimated at 10 Mbps).
+See also: https://github.com/waku-org/research/issues/31
+
+# Store protocol
+
+We first focus on i13n of the Store protocol.
+Similar techniques may be later applied to other protocols.
+
+Store is a client-server protocol.
+A client asks the node to respond with relevant messages previously relayed through the Relay protocol.
+A relevant message is a message that has been broadcast via Relay within the specified time frame.
+The response may be split into multiple parts, as specified by pagination parameters.
+
+TODO: Strictly speaking, the definition of relevant is inconsistent because there is no consensus over messages. A message may be broadcast but not received by some nodes. Does this happen often? Can and should we do something about it?
+
+## Desired behavior
+
+The desired behavior of a Store-server node is to store all non-ephemeral messages forever.
+
+TODO: address the obvious concern that storing everything forever is unsustainable. Should there be some cut-off time after which old messages are no longer stored?
+
+Let's say, a client issues a request to the server.
+We want the following to happen:
+
+- the server responds quickly;
```

Comment:
> all the messages in the response are relevant;
the response contains only relevant messages.

we want this to happen but unsure how to enforce it. perhaps we can just assume that its not the case, and use multiple nodes to define what relevant is. eg, if i query 10 nodes and 7 report the same message, then its "relevant".

at first this might seem very bandwidth intensive, but we can play with hashes, trees and other tricks to not naively request the whole messages to 10 nodes. cc @ABresting

Reply:
That is a meta-question: should we define the ideal behavior even if it's unattainable? I tend to think we should. I would _define_ "relevant" for the ideal case, and then separately discuss what approximations are feasible in practice. "Assuming probability `p` that a random server is malicious, and probability `q` that a message is lost, querying `n` servers gives this probability that..." etc.